SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp

```diff
@@ -0,0 +1,340 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "VisibilityTest.h"
+#include "ReplicatedVisibilityTestActor.h"
+#include "GameFramework/PlayerController.h"
+#include "SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMovement/TestMovementCharacter.h"
+#include "GameFramework/CharacterMovementComponent.h"
+#include "Kismet/GameplayStatics.h"
+#include "Net/UnrealNetwork.h"
+#include "Containers/Array.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialGDKSettings.h"
+
+/**
+ * This test tests if a bHidden Actor is replicating properly to Server and Clients.
+ * 
+ * The test includes a single server and two client workers.
+ * The flow is as follows:
+ *  - Setup:
+ *	  - On cube actor already placed in the level Location FVector(0.0f, 0.0f, 50.0f)
+ *    - Two test character pawns are spawned, one for each client
+ *    - The controllers for each client  possess the spawned test character pawns
+ *  - Test:
+ *    - Server moves one Client to a remote position where Client can not see the HiddenActor(Cube) .
+ *	  - Check if the Clients can see the HiddenActor(Cube).
+ *	  - Server sets the HiddenActor(Cube) as bHidden.
+ *	  - Server moves the remote Client back close to the HiddenActor(Cube).
+ *	  - Check if the Clients can see the HiddenActor(Cube).
+ *	  - Server sets the HiddenActor(Cube) as Visible (bHidden = false).
+ *	  - Check if the Clients can see the HiddenActor(Cube).
+ *  - Cleanup:
+ *    - The clients repossess their default pawns
+ *    - The test pawns are destroyed
+ */
+
+AVisibilityTest::AVisibilityTest()
+	: Super()
+{
+	Author = "Evi";
+	Description = TEXT("Test Actor Visibility");
+}
+void AVisibilityTest::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(AVisibilityTest, Controllers);
+	DOREPLIFETIME(AVisibilityTest, TestPawns);
+}
+
+void AVisibilityTest::BeginPlay()
+{
+	Super::BeginPlay();
+
+	CharacterRemoteLocation = FVector(20000.0f, 20000.0f, 50.0f);
+	Character1StartingLocation = FVector(0.0f, 120.0f, 50.0f);
+	Character2StartingLocation = FVector(0.0f, 240.0f, 50.0f);
+
+	{	// Step 1 - Set up the world Spawn PlayerCharacter and make sure the cube is in the world.
+		AddStep(TEXT("ServerSetup"), FWorkerDefinition::Server(1), nullptr, [this](ASpatialFunctionalTest* NetTest) {
+			AVisibilityTest* Test = Cast<AVisibilityTest>(NetTest);
```

Comment:
My personal preference is that if you capture `this `in the lambda declaration above, you don't need to cast the `NetTest `as a `AVisibilityTest`*, and instead of using `Test->TestPawns.Empty()` you can just use `TestPawns.Empty()`, and the same applies for every call of the type `Test->Something` or `NetTest->Something` , they can be replaced by  `Something` if you wish. It's not important, just a way of making the code look a bit better in my opinion.